File: src/main/java/editor/database/attributes/Loyalty.java
package editor.database.attributes;

import editor.database.card.Card;
import editor.util.Parsers;

/**
 * This class represents the starting loyalty of a planeswalker.  All {@link Card}s have them, but they
 * are invisible for non-planeswalkers. A value of {@link Double#NaN} means there is no loyalty, -1 means
 * X loyalty (set by something else such as X in mana cost), and -2 means variable loyalty (*).
 *
 * @param value numerical value of the starting loyalty
 * 
 * @author Alec Roelke
 */
public record Loyalty(double value) implements OptionalAttribute, Comparable<Loyalty>
{
    /**
     * Constant for an arbitrary card with no loyalty.
     */
    public static final Loyalty NO_LOYALTY = new Loyalty(Double.NaN);

    private static final double X = -1;
    private static final double STAR = -2;

    /**
     * Create a new Loyalty with the given value. Loyalty creatd this way can't vary.
     *
     * @param v value of the new Loyalty
     */
    public Loyalty(int v)
    {
        this((double)Math.max(0, v));
    }

    /**
     * Parse a String for a loyalty value.  It can either be a number, "X," or "*,"
     * and if anything else is used it is considered to not exist.
     *
     * @param s
     */
    public Loyalty(String s)
    {
        this (switch (s.toUpperCase()) {
            case "X" -> X;
            case "*" -> STAR;
            case ""  -> Double.NaN;
            default  -> Parsers.tryParseDouble(s).orElse(Double.NaN);
        });
    }

    @Override
    public int compareTo(Loyalty other)
    {
        return Double.compare(value, other.value);
    }

    @Override
    public boolean exists()
    {
        return !Double.isNaN(value);
    }

    @Override
    public String toString()
    {
        if (value == X)
            return "X";
        else if (value == STAR)
            return "*";
        else if (Double.isNaN(value))
            return "";
        else
            return Integer.toString((int)value);
    }

    /**
     * @return <code>true</code> if this Loyalty is variable (X, *), and <code>false</code> otherwise.
     */
    public boolean variable()
    {
        return value < 0;
    }
}

File: src/main/scala/editor/database/attributes/Loyalty.scala
package editor.database.attributes

/**
 * Starting loyalty of a planeswalker card. Some planeswalkers have variable loyalties,
 * either starting (X) or based on game state (*), instead of concrete numbers.
 * 
 * @constructor create a new starting loyalty
 * @param value amount of loyalty
 * 
 * @author Alec Roelke
 */
case class Loyalty(value: Double) extends Ordered[Loyalty] {
  /** Whether or not the card actually has loyalty (i.e. is a planeswalker) */
  val exists = !value.isNaN

  /** Whether or not the starting loyalty is variable. */
  val variable = value < 0

  override def compare(that: Loyalty) = {
    val c = value.compare(that.value)
    if (c == 0) {
      if (variable && !that.variable)
        1
      else if (!variable && that.variable)
        -1
      else // both variable or neither variable
        0
    } else c
  }

  override val toString = value match {
    case Loyalty.X    => "X"
    case Loyalty.STAR => "*"
    case _ => value.toInt.toString
  }
}

object Loyalty {
  /**
   * Create a loyalty value from a string. The only permitted options are integers,
   * "X", "*", and the empty string (to represent no loyalty).
   * 
   * @param s string to parse
   * @return a new loyalty value
   */
  def apply(s: String): Loyalty = Loyalty(s.toUpperCase match {
    case "X" => Loyalty.X
    case "*" => Loyalty.STAR
    case  u  => u.toDouble
  })

  /** Number that represents a starting loyalty of X (determined as the planeswalker enters). */
  val X: Double = -1
  /** Number that represents continuously variable loyalty (based on some other part of the game state). */
  val STAR: Double = -2
}